python/ray/data/tests/test_numpy_support.py

```diff
@@ -0,0 +1,129 @@
+import numpy as np
+import torch
+import pytest
+
+import ray
+from ray.data.tests.conftest import *  # noqa
+from ray.tests.conftest import *  # noqa
+
+
+class UserObj:
+    def __eq__(self, other):
+        return isinstance(other, UserObj)
+
+
+def do_map_batches(data):
+    ds = ray.data.range(1)
+    ds = ds.map_batches(lambda x: {"output": data})
+    return ds.take_batch()["output"]
+
+
+def assert_structure_equals(a, b):
+    assert type(a) == type(b), (type(a), type(b))
+    assert type(a[0]) == type(b[0]), (type(a[0]), type(b[0]))  # noqa: E721
+    assert a.dtype == b.dtype
+    assert a.shape == b.shape
+    for i in range(len(a)):
+        assert np.array_equiv(a[i], b[i])
+
+
+def test_list_of_scalars(ray_start_regular_shared):
+    data = [1, 2, 3]
+    output = do_map_batches(data)
+    assert_structure_equals(output, np.array([1, 2, 3], dtype=np.int64))
+
+
+def test_list_of_numpy_scalars(ray_start_regular_shared):
+    data = [np.int64(1), np.int64(2), np.int64(3)]
+    output = do_map_batches(data)
+    assert_structure_equals(output, np.array([1, 2, 3], dtype=np.int64))
+
+
+def test_list_of_objects(ray_start_regular_shared):
+    data = [1, 2, 3, UserObj()]
+    output = do_map_batches(data)
+    assert_structure_equals(output, np.array([1, 2, 3, UserObj()]))
```

Comment:
check that this is object dtype right?

Reply:
Yup, assert struct equals will check the types are exactly equal. The literal on the right returns an array with object dtype.